libgdf/src/join/sort/sort-join.cuh

```diff
@@ -1,246 +1,227 @@
-    int join_count;
-    mem_t<int> scanned_sizes = scan_join_bounds(bounds, a_count, b_count, context, false,
-                                                join_count);
-    auto output = compute_joined_indices(bounds, scanned_sizes, a_count,
-                                               join_count, context, false, 0);
-    return output;
+template<JoinType join_type,
+    typename column_type,
+    typename index_type>
+gdf_error sort_join_typed(
+        gdf_column * const output_l,
+        gdf_column * const output_r,
+        gdf_column * const leftcol,
+        gdf_column * const rightcol,
+        bool flip_results = false) {
+    JoinBounds<index_type> bounds =
+        compute_join_bounds<column_type, index_type>(
+                static_cast<column_type*>(leftcol->data), leftcol->size,
+                static_cast<column_type*>(rightcol->data), rightcol->size);
+    thrust::device_vector<index_type> scanned_sizes =
+        scan_join_bounds<join_type, index_type>(bounds);
+    std::pair<gdf_column, gdf_column> join_result =
+        compute_joined_indices<join_type, index_type>(bounds, leftcol, rightcol, scanned_sizes);
+    *output_l = join_result.first;
+    *output_r = join_result.second;
+    if (flip_results) {
+        *output_l = join_result.first;
+        *output_r = join_result.second;
+    }
+    return GDF_SUCCESS;
 }
 
-template<typename launch_arg_t = empty_t,
-  typename a_it, typename b_it, typename comp_t>
-std::pair<gdf_column, gdf_column>
-full_join(a_it a, int a_count, b_it b, int b_count,
-           comp_t comp, context_t& context)
-{
-    _join_bounds bounds = compute_join_bounds(a, a_count, b, b_count, comp,
-                                              context);
-    int join_count;
-    mem_t<int> scanned_sizes = scan_join_bounds(bounds, a_count, b_count, context, false,
-                                                join_count);
-    int append_count;
-    mem_t<int> matches = full_join_count_matches(a, a_count, b, b_count,
-                                                  comp, context, append_count );
-    auto output = compute_joined_indices(bounds, scanned_sizes, a_count,
-                                               join_count, context, false, append_count);
-    full_join_append_right(
-            static_cast<int*>(output.first.data),
-            static_cast<int*>(output.second.data),
-            matches, append_count, join_count, context);
-    return output;
+template<JoinType join_type,
+    typename index_type>
+gdf_error compute_sort_join(
+        gdf_column * const output_l,
+        gdf_column * const output_r,
+        gdf_column * const lcol,
+        gdf_column * const rcol,
+        bool flip = false) {
+    int column_width_bytes{0};
+    gdf_error gdf_status = get_column_byte_width(lcol, &column_width_bytes);
+
+    if(GDF_SUCCESS != gdf_status)
+        return gdf_status;
+
+    switch (column_width_bytes) {
```

Comment:
This `switch` is problematic. 

By casting everything to an `int*_t`, this will technically produce incorrect orderings for unsigned ints as well as floating point values when there is a mix of positive/negative values, as well as sentinel values such as `NaN`. However, I believe for the purposes of this algorithm this is technically _fine_ because you're just trying to group together bit-wise equal keys, and by casting them all to `int` will do this. 

However, where this breaks down is when the type of the column isn't a primitive type (e.g., GDF_STRING) in which case an overloaded `operator<` would need to be used.

In short, I would say this `switch` should be updated to dispatch the correct underlying type, and not just the byte-width. The `gdf_type_dispatcher` I've created should be used for this (though it is not yet merged). See here: https://github.com/rapidsai/cudf/blob/fb8163df529574e3c75603a5e3c14d3164dc3245/libgdf/src/gdf_type_dispatcher.cuh

Reply:
Since the dispatcher has not been merged yet I have reverted to the old style switch case dependent on the column type. Perhaps this function can be changed once it is in master.

Reply:
The dispatcher is merged.